### DISCUSSION TITLE
Custom Provider: Cannot find module '@ioc:*' or its corresponding type declarations

### DISCUSSION THREAD

**Discussion by LeCoupa** [adonisjs/core#2040] [General]
This is probably something easy to solve but I can't figure out why TypeScript keep saying this when I import my custom provider:

> Cannot find module '@ioc:Elastic' or its corresponding type declarations

<img width="1014" alt="Screenshot 2020-12-12 at 17 34 16" src="https://user-images.githubusercontent.com/1658644/101989390-454c6300-3ca0-11eb-8bc1-7582e136a1ed.png">

Here is my custom provider:

```javascript
// NPM
import { IocContract } from '@adonisjs/fold'
import { Client as ElasticClient } from '@elastic/elasticsearch'

// Contracts
import { ElasticBodyContract, SortObjectContract } from 'contracts/elastic'

export default class ElasticProvider {
  constructor(protected container: IocContract) {}

  public async register() {
    const Env = (await import('@ioc:Adonis/Core/Env')).default

    this.container.singleton('Elastic', () => {
      const elasticClient = new ElasticClient({
        node: Env.get('ES_URL'),
      })

      return elasticClient
    })
  }
}

```

**Comment by steschwa** [comment]
Typescript does not know about the IOC-Bindings at **compile time** (because they are evaluated at **runtime**). Therefore you have to create a types declaration file for your custom service like so:

**contracts/WhateverNameYouWant.d.ts**
```ts 
declare module "@ioc:Elastic" {
    import { Client as ElasticClient } from "@elastic/elasticsearch"
    declare const elasticClient: ElasticClient
    export default elasticClient
}
```

Under the hood (in their corresponding packages) Adonis does exactly that for built-in providers.

**Reply by steschwa** [reply]
Theoretically you can rename the declaration file extension from **.d.ts** to only **.ts** but i personally prefer to use **.d.ts** to know that only type declarations are within this file and no other logic

**Reply by LeCoupa** [reply]
Thank you :) I also updated all my contracts to .d.ts 👍🏻

**Comment by hossain45** [comment]
in my case, 
`      "@ioc/*": ["./node_modules/@adonisjs/core/build/standalone/ioc/*"]
`
was missing from my tsconfig file

here's the full config:
`{`
`  "extends": "./node_modules/adonis-preset-ts/tsconfig",`
`  "include": ["**/*"],`
`  "exclude": ["node_modules", "build"],`
`  "compilerOptions": {`
`   "outDir": "build",`
`    "rootDir": "./",`
`    "sourceMap": true,`
`    "paths": {`
`      "App/*": ["./app/*"],`
`      "Config/*": ["./config/*"],`
`      "Contracts/*": ["./contracts/*"],`
`      "Database/*": ["./database/*"],`
`      "@ioc/*": ["./node_modules/@adonisjs/core/build/standalone/ioc/*"]`
`    },`
`    "types": [`
`      "@adonisjs/core",`
`      "@adonisjs/repl",`
`      "@adonisjs/lucid",`
`      "@adonisjs/auth",`
`      "@adonisjs/shield",`
`      "@adonisjs/mail"`
`    ],`
`  }`
`}`
`

**Reply by RomainLanz** [reply]
The code you added seems incorrect.

`@ioc:` imports were part of AdonisJS 5's IoC Container system and aren't needed in AdonisJS 6.

I'm not entirely sure what problem you're trying to solve. If you need further help, consider opening a new discussion.

**Reply by hossain45** [reply]
you are right

I have not migrated to AdonisJS6 yet.
